packages/nouns-contracts/contracts/StreamEscrow.sol

```diff
@@ -0,0 +1,307 @@
+        if (block.timestamp < lastForwardTimestamp + 24 hours) {
+            return;
+        }
+
+        lastForwardTimestamp = toUint48(block.timestamp);
+
+        uint256 ethStreamedPerTickBefore = ethStreamedPerTick;
+        sendETHToTreasury(ethStreamedPerTickBefore);
+
+        uint32 newTick = increaseTicksAndFinishStreams();
+
+        emit StreamsForwarded(newTick, ethStreamedPerTickBefore, ethStreamedPerTick, lastForwardTimestamp);
+    }
+
+    /**
+     * @notice Cancels multiple streams at once.
+     * @param nounIds The IDs of the Noun tokens to cancel streams for.
+     */
+    function cancelStreams(uint256[] calldata nounIds) external {
+        for (uint256 i; i < nounIds.length; ++i) {
+            cancelStream(nounIds[i]);
+        }
+    }
+
+    /**
+     * @notice Cancels a stream for a specific Noun token. Transfers the Noun to `nounRecipient`
+     *  and refunds the remaining ETH back to the caller.
+     * @notice The caller must be the Noun owner.
+     * @param nounId The ID of the Noun token to cancel the stream for.
+     */
+    function cancelStream(uint256 nounId) public {
+        require(isStreamActive(nounId), 'stream not active');
+
+        // transfer noun to treasury
+        nounsToken.transferFrom(msg.sender, nounsRecipient, nounId);
+
+        // cancel stream
+        streams[nounId].canceled = true;
+        Stream memory stream = streams[nounId];
+        ethStreamedPerTick -= stream.ethPerTick;
+        ethStreamEndingAtTick[stream.lastTick] -= stream.ethPerTick;
+
+        // calculate how much needs to be refunded
+        uint256 ticksLeft = stream.lastTick - currentTick;
+        uint256 amountToRefund = stream.ethPerTick * ticksLeft;
+        (bool sent, ) = msg.sender.call{ value: amountToRefund }('');
+        require(sent, 'failed to send eth');
+
+        emit StreamCanceled(nounId, amountToRefund);
+    }
+
+    /**
+     * @notice Fast-forwards a stream by a certain number of ticks.
+     * @param nounId The ID of the Noun token to fast-forward the stream for.
+     * @param ticksToForward The number of ticks to fast-forward the stream by.
+     * @custom:throws 'ticksToForward must be positive' if `ticksToForward` is not positive.
+     * @custom:throws 'ticksToFoward too large' if `ticksToForward` is larger than the remaining ticks.
+     * @custom:throws 'not noun owner' if the caller is not the Noun owner.
+     * @custom:throws 'stream not active' if the stream is not active.
+     */
+    function fastForwardStream(uint256 nounId, uint32 ticksToForward) public {
+        require(ticksToForward > 0, 'ticksToForward must be positive');
+        require(nounsToken.ownerOf(nounId) == msg.sender, 'not noun owner');
+
+        Stream memory stream = streams[nounId];
+        uint32 currentTick_ = currentTick;
+        require(isStreamActive(stream, currentTick_), 'stream not active');
+        uint32 lastTick = stream.lastTick;
```

Comment:
wondering why you chose to copy this memory `lastTick` to a stack var?

Reply:
probably gas optimizations leftover. I removed it. a2ab0a951ecb90159f1e347e9d207ae1cc8d61f5